zap_text_encoder.go

```diff
@@ -0,0 +1,608 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package log
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"math"
+	"strings"
+	"sync"
+	"time"
+	"unicode/utf8"
+
+	"go.uber.org/zap/buffer"
+	"go.uber.org/zap/zapcore"
+)
+
+// DefaultTimeEncoder serializes time.Time to a human-readable formatted string
+func DefaultTimeEncoder(t time.Time, enc zapcore.PrimitiveArrayEncoder) {
+	s := t.Format("2006/01/02 15:04:05.000 -07:00")
+	if e, ok := enc.(*textEncoder); ok {
+		for _, c := range []byte(s) {
+			e.buf.AppendByte(c)
+		}
+		return
+	}
+	enc.AppendString(s)
+}
+
+// ShortCallerEncoder serializes a caller in file:line format.
+func ShortCallerEncoder(caller zapcore.EntryCaller, enc zapcore.PrimitiveArrayEncoder) {
+	enc.AppendString(getCallerString(caller))
+}
+
+func getCallerString(ec zapcore.EntryCaller) string {
+	if !ec.Defined {
+		return "<unknown>"
+	}
+
+	idx := strings.LastIndexByte(ec.File, '/')
+	buf := _pool.Get()
+	for i := idx + 1; i < len(ec.File); i++ {
+		b := ec.File[i]
+		switch {
+		case b >= 'A' && b <= 'Z':
+			buf.AppendByte(b)
+		case b >= 'a' && b <= 'z':
+			buf.AppendByte(b)
+		case b >= '0' && b <= '9':
+			buf.AppendByte(b)
+		case b == '.' || b == '-' || b == '_':
+			buf.AppendByte(b)
+		default:
+		}
+	}
+	buf.AppendByte(':')
+	buf.AppendInt(int64(ec.Line))
+	caller := buf.String()
+	buf.Free()
+	return caller
+}
+
+// For JSON-escaping; see textEncoder.safeAddString below.
+const _hex = "0123456789abcdef"
+
+var _textPool = sync.Pool{New: func() interface{} {
+	return &textEncoder{}
+}}
+
+var (
+	_pool = buffer.NewPool()
+	// Get retrieves a buffer from the pool, creating one if necessary.
+	Get = _pool.Get
+)
+
+func getTextEncoder() *textEncoder {
```

Comment:
why do we need a pool for the encoder?

Reply:
Sometimes we need `cloned` to create a temp encoder, so use a pool to reduce allocation.